src/policy/sft_map.rs

```diff
@@ -0,0 +1,506 @@
+use super::sft::*;
+use crate::util::metadata::side_metadata::SideMetadataSpec;
+use crate::util::Address;
+#[cfg(debug_assertions)]
+use crate::util::ObjectReference;
+#[cfg(debug_assertions)]
+use crate::vm::VMBinding;
+
+/// SFTMap manages the SFT table, and mapping between addresses with indices in the table. The trait allows
+/// us to have multiple implementations of the SFT table.
+pub trait SFTMap {
+    /// Check if the address has an SFT entry for it (including an empty SFT entry). This is mostly a bound check
+    /// to make sure that we won't have an index-out-of-bound error. For the sake of performance, the implementation
+    /// of other methods in this trait (such as get(), update() and clear()) does not need to do this check implicitly.
+    /// Instead, they assume the address has a valid entry in the SFT. If an address could be arbitary, they should call this
+    /// method as a pre-check before they call any other methods in the trait.
+    fn has_sft_entry(&self, addr: Address) -> bool;
+
+    /// Get the side metadata spec this SFT map uses.
+    fn get_side_metadata(&self) -> Option<&SideMetadataSpec>;
+
+    /// Get SFT for the address. The address must have a valid SFT entry in the table.
+    fn get(&self, address: Address) -> &dyn SFT;
```

Comment:
Yeah, that is possible. But that means whenever we try to get SFT for an address, we need the extra check. I feel that may not be always desirable, as in most cases, we know we are getting SFT for an object (which is guaranteed in our spaces). I can change `get()` into `get_checked()` and `get_unchecked()`: the `checked()` version will do the check you suggested, and will return `EMPTY_SPACE_SFT` if the address is out-of-bounds, and the `unchecked()` version will simply index into the array (we could also use [`get_unchecked`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get_unchecked)  from Rust `Vec`).

Reply:
I have done this change in https://github.com/mmtk/mmtk-core/pull/632/commits/2ed65b0c81c6813c7d62ecdb3e29bd18b7a96221. Let me know if there is anything alarming. Otherwise, I will treat this PR as approved.